llvm/lib/Transforms/Utils/SimplifyIndVar.cpp

```diff
@@ -644,6 +644,13 @@ bool SimplifyIndvar::replaceIVUserWithLoopInvariant(Instruction *I) {
 
   auto *Invariant = Rewriter.expandCodeFor(S, I->getType(), IP);
 
+  if (!LI->replacementPreservesLCSSAForm(I, Invariant)) {
+    LLVM_DEBUG(dbgs() << "INDVARS: Can not replace IV user: " << *I
+                      << " with loop invariant: " << *S
+                      << " as it breaks LCSSA form " << '\n');
+    return false;
+  }
```

Comment:
SCEVExpander is also supposed to preserve LCSSA form (if PreserveLCSSA is enabled, which it is by default), so I think the actual fix needs to be in SCEVExpander. It's not possible to abort a transform after using SCEVExpander.

Reply:
There's call to `fixupLCSSAFormFor` in SCEVExpander::expand, which in turn calls formLCSSAForInstructions. So, I am not sure why it's failing to insert a LCSSA Phi

Reply:
@nikic it seems the LI is not being updated, because UseLoop and DefLoop can't be equal that's for sure. The InsertionPt and DefI are not in the same loop !

Reply:
@nikic I studied the code, I don't think SCEVExpander is the right place. As we don't pass the Instruction to be replaced by the expanded value to the SCEVExpander at any point. Either we need a expanded value which doesn't make changes to the IR so that we can safely check for the LCSSA, or we need to somehow pass Instruction to be replaced to SCEVExpander and then do the check inside ? (I don't think we should do this).

What do you think, would be the ideal way to make this change ?

Reply:
Okay, I get the problem now. SCEVExpander insertion point is the loop-invariant location, but the actual use will be inside an exit block.

I think the options to fix this would be:
1. Make a check that LCSSA is not broken before calling Rewriter, under the assumption that the instruction will be inserted at the insertion point. (This may have some false positives if instruction reuse occurs, but I think it's fine.)
2. Fix up LCSSA form in this special case after the transform. So if `!replacementPreservesLCSSAForm` insert a phi node.